TESTFLIGHT nodes inside MEC CONFIG nodes

lpgagnon opened the pull request:
Put TESTFLIGHT nodes inside CONFIG nodes to avoid having to maintain them separately and keep them in sync; move them out to the PART top-level just before they get processed.

Main motivation: avoid having to repeat the same 'copy rated burn time to CONFIG' lines in every TESTFLIGHT node, removing risk of copy-paste errors (naming the wrong config, etc). ([like this from today](https://discord.com/channels/319857228905447436/620690446540341261/1027167101372338196))

Likewise would avoid copy-paste mismatches between CONFIG node name and TESTFLIGHT node name

Might also help load times? BEFORE:zTestFlight is weirdly slow, and the ratedburntime copying is the only fancy thing it does. (this is fancier, but differently so)

One argument against: somebody probably likes the 'all the TF configs at the bottom of the file' pattern.

initial mergeable submission: 1 commit for the plumbing, 1 for a first set of engines

NathanKell commented:
Oh hell yes.

Capkirk123 commented:
This is pretty nice, would probably help clean things up a bit.

al2me6 commented on GameData/RealismOverhaul/Engine_Configs/TinyTim_Config.cfg:
```diff
@@ -166,43 +179,70 @@
 }
 
-@PART[*]:HAS[@MODULE[ModuleEngineConfigs]:HAS[@CONFIG[30Klbf]],!MODULE[TestFlightInterop]]:BEFORE[zTestFlight]:NEEDS[TestLite|TestFlight]
+// TODO: move elsewhere. probably wherever FOR[zTestFlight] lives.
```
`TestFlight_Generic_Engines.cfg` is probably good.

al2me6 commented on GameData/RealismOverhaul/Engine_Configs/TinyTim_Config.cfg:
```diff
@@ -166,43 +179,70 @@
+				&testedBurnTime = dummy
+			}
+		}
+		@CONFIG_wip
+		{
+			%ratedBurnTime = #$TESTFLIGHT/ratedBurnTime$
+			%ratedContinuousBurnTime = #$TESTFLIGHT/ratedContinuousBurnTime$
+			// TODO: maybe something like this to display testedBurnTime SOMEWHERE in-game
+			// @description = #$description$ testedBurnTime=$TESTFLIGHT/testedBurnTime$
+			// @description ^= : testedBurnTime=dummy$::
+		}
+		@CONFIG_wip:HAS[#ratedBurnTime[dummy]] { !ratedBurnTime = delete }
+		@CONFIG_wip:HAS[#ratedContinuousBurnTime[dummy]] { !ratedContinuousBurnTime = delete }
```
Maybe

```
@CONFIG_wip:HAS[@TESTFLIGHT:HAS[#ratedBurnTime]] { &ratedBurnTime = #$TESTFLIGHT/ratedBurnTime$ }
// &c.
```

lpgagnon replied:
I think I wanted to avoid needing multiple nested lookups, but I didn't end up in a good place. Will use this.
(and then maybe later when it's all done, see if there's a perf difference)

al2me6 commented:
Third.

lpgagnon commented:
updated: cleaned up the plumbing (thanks al2me6) and split into its own commit; convert configs from 18KS7800 to Agena (plus template and tim) in separate commit. github diffs are almost meaningful, but not quite

plausibly mergeable as-is, with the rest of the engines to be updated incrementally over time. I plan on continuing to work my way down the list, but with no particular hurry. might make sense to merge now, might make sense to wait until there's more of a critical mass. Probably doesn't make sense to wait until everything's done, because 1) I might get bored 2) the longer it waits in limbo, the more likely conflicts get, the more likely I just give up entirely

if someone feels like chipping in, sync up on discord.

al2me6 commented:
Hm. It just occurred to me that a neat side effect of this change is that it becomes a lot easier for external tools to parse configs. (One day I'll get back to working on Engine Config Browser...)

StonesmileGit commented on GameData/RealismOverhaul/Engine_Configs/Agena_XLR81_Config.cfg:
```diff
@@ -304,6 +319,23 @@
+				//Identical to XLR81-BA-5 other than nozzle, give same reliability since only two were launched
+				ignitionReliabilityStart = 0.889167
+				ignitionReliabilityEnd = 0.982500
+				
```
Auto-generated potential formatting fix:
```suggestion

```

StonesmileGit commented on GameData/RealismOverhaul/Engine_Configs/Agena_XLR81_Config.cfg:
```diff
@@ -338,6 +370,24 @@
+				safeOverburn = true
+				ignitionReliabilityStart = 0.889167
+				ignitionReliabilityEnd = 0.982500
+				
```
Auto-generated potential formatting fix:
```suggestion

```

lpgagnon commented:
updated: added support for `clusterMultiplier` to the pre-post-processing steps, and converted all engines up to L. (also squashed the bot's cleanup)

lpgagnon commented:
updated: engines from M to Z converted, this is now complete.

hypothesis confirmed: this change makes the BEFORE[zTestFlight] pass go from ~90s to ~5s on my machine (total MM time goes from ~5:45 to ~4:15)

DRVeyl commented on GameData/RealismOverhaul/TestFlight_Generic_Engines.cfg:
```diff
@@ -1,3 +1,55 @@
+		{
+			|_ = CONFIG_wip // for the MM_PATCH_LOOP below
+			@TESTFLIGHT,* {
+				name = #$../name$
```
Should be `%name = `?

lpgagnon replied:
probably a good idea. the intent was to never provide a name in the TESTFLIGHT nodes themselves, but there's a pretty good chance I missed deleting some.

Capkirk123 replied:
This doesn't appear to have actually been done? I agree that it should be just to safeguard against any poorly written configs.

lpgagnon commented on GameData/RealismOverhaul/Engine_Configs/18KS7800_Config.cfg:
```diff
@@ -118,23 +118,18 @@
 				key = 0.10 0.250
 				key = 0.00 0.040
 			}
+			TESTFLIGHT:NEEDS[TestLite,TestFlight]
```
these come straight from the originals. if it's an AND, testflight configs have been broken since testlite was added, and I don't think that's the case. (and obviously I tested this at least a little bit)

lpgagnon replied:
... but I do see the mix in the originals. I'll see how quick it is to fix

lpgagnon replied:
Fixed (and resolved rd-0122 conflict). Turns out it was just those 2 SRBs, the very first files I converted. must have copied the NEEDS from the post-processing file (which [correctly] uses `,`)

Capkirk123 commented:
Ok, I went to test this, and engines don't actually fail? Ignition failures seem to work fine, but all other failures (performance loss, thrust loss, shutdown, explosion) don't seem to work, the TF window and PAW say the failure has occurred but the engine keeps functioning normally.
I don't really understand the MM magic in testflight_generic_engines, @al2me6 any ideas?

Capkirk123 commented on GameData/RealismOverhaul/TestFlight_Generic_Engines.cfg:
```diff
@@ -1,3 +1,55 @@
+		{
+			|_ = CONFIG_wip // for the MM_PATCH_LOOP below
+			@TESTFLIGHT,* {
+				name = #$../name$
```
As DRVeyl said, this should have a ``%``.

lpgagnon replied:
Not sure if I only imagined fixing it or somehow lost the change. In any case, re-done; also rebased.Not sure if I only imagined fixing it or somehow lost the change. In any case, re-done; also rebased.

Capkirk123 reviewed (changes requested):
Fixed my KSP install and tested again. Everything appears to work as intended, and I can confirm patching goes much faster with these changes. Just needs a ``%`` and should be good to go.

Capkirk123 reviewed (approved):
Looks good!

StonesmileGit commented on GameData/RealismOverhaul/Engine_Configs/RSRM_Config.cfg:
```diff
@@ -271,6 +282,14 @@
+			}
 		}
 
 		+CONFIG[RSRM-1981]
```
With the TF data now living inside the config, the correct values are in the '86 config, so better to copy from there.
With the current copy, the techTransfer key doesn't exist, so no edit is made
```suggestion
		+CONFIG[RSRM-1986]
```

lpgagnon replied:
I'm almost sure that's what I had intended to do. Accepting suggestion; will likely squash later.

Capkirk123 commented on GameData/RealismOverhaul/TestFlight_Generic_Engines.cfg:
```diff
@@ -1,3 +1,55 @@
+// this lets us copy some fields from one to the other, but then the
+// FOR:zTestFlight processing steps need them directly in the PART node.
+// Must use BEFORE to preserve TestLite compatibility.
+@PART[*]:HAS[@MODULE[ModuleEngineConfigs]]:BEFORE[zTestFlight]
```
All instances of ModuleEngineConfigs should be replaced by Module*EngineConfigs so ModuleBimodalEngineConfigs also gets included in the patching.

lpgagnon replied:
done; and confirmed LR129 has a working TF config again

lpgagnon commented on GameData/RealismOverhaul/Engine_Configs/R4D11_Config.cfg:
```diff
@@ -234,25 +234,16 @@
-		ignitionDynPresFailMultiplier = 0.1
-		cycleReliabilityStart = 0.98
-		cycleReliabilityEnd = 0.9985
+			TESTFLIGHT:NEEDS[TestLite|TestFlight]
```
fixed
